cpp/inc/bond/core/nullable.h

```diff
@@ -37,308 +37,303 @@ use_value
 
     /// @brief Default constructor
-    nullable()
-        : _value()
+    nullable() BOND_NOEXCEPT_IF(
+        std::is_nothrow_default_constructible<alloc_holder>::value
+        && std::is_nothrow_default_constructible<pointer>::value)
+        : alloc_holder(),
+          _value()
     {}
 
     /// @brief Construct nullable using specified allocator instance
     explicit
     nullable(const allocator_type& alloc)
-        : Allocator(alloc),
+        : alloc_holder(alloc),
           _value()
     {}
 
+    /// @brief Copy constructor
+    nullable(const nullable& other)
+        : alloc_holder(other),
+          _value(other.hasvalue() ? new_value(other.value()) : pointer())
+    {}
+
+    nullable(const nullable& other, const allocator_type& alloc)
+        : alloc_holder(alloc),
+          _value(other.hasvalue() ? new_value(other.value(), alloc) : pointer())
+    {}
+
+    nullable(nullable&& other) BOND_NOEXCEPT_IF(
+            std::is_nothrow_move_constructible<alloc_holder>::value
+            && std::is_nothrow_move_constructible<pointer>::value
+            && BOND_NOEXCEPT(other._value = {}))
+        : alloc_holder(std::move(other)),
+          _value(std::move(other._value))
+    {
+        other._value = {};
+    }
+
+    nullable(nullable&& other, const allocator_type& alloc)
+        : alloc_holder(alloc),
+          _value(other.alloc_holder::get() == alloc
+              ? std::move(other._value)
+              : (other.hasvalue() ? new_value(std::move(*other._value), alloc) : pointer()))
+    {
+        other._value = {};
+    }
+
     /// @brief Construct from an instance T
     explicit
-    nullable(const value_type& value,
-             const allocator_type& alloc = allocator_type())
-        : Allocator(alloc),
+    nullable(const T& value, const allocator_type& alloc = {})
+        : alloc_holder(alloc),
           _value(new_value(value))
     {}
 
-    /// @brief Copy constructor
-    nullable(const nullable& src)
-        : Allocator(src.base()),
-          _value(src.hasvalue() ? new_value(src.value()) : real_pointer())
+    explicit
+    nullable(T&& value, const allocator_type& alloc = {})
+        : alloc_holder(alloc),
+          _value(new_value(std::move(value)))
     {}
 
-    nullable(const nullable& src, const allocator_type& alloc)
-        : Allocator(alloc),
-          _value(src.hasvalue() ? new_value(src.value(), alloc) : real_pointer())
-    {}
+    /// @brief Assignment operator
+    nullable& operator=(const nullable& other)
+    {
+        nullable(other).swap(*this);
+        return *this;
+    }
+
+    nullable& operator=(nullable&& other)
+    {
+        nullable(std::move(other)).swap(*this);
+        return *this;
+    }
 
     ~nullable()
     {
         reset();
     }
 
-    /// @brief Assignment operator
-    nullable& operator=(const nullable& src)
+    bool hasvalue() const BOND_NOEXCEPT
     {
-        nullable(src).swap(*this);
-        return *this;
+        return !!_value;
```

Comment:
Is the `!!` idiom to handle custom `pointer` types that don't have an `explicit operator bool`?

Reply:
Old style to enforce 0/1 :) Will make an explicit cast.